### PR TITLE
ADR: Remote-only packaging of MLMD Python lib

### PR DESCRIPTION
**Pull request by tarilabs**
## Description
ADR for Remote-only packaging of MLMD Python lib as previously discussed.

## How Has This Been Tested?
n/a

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [n/a] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [n/a] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+Pros:
+
+* Only one artifact to publish on PyPI
+
+Cons:
+
+* Google code (copyright header) is mixed with Model Registry code
+    * at this stage is not clear if any implications with uplifting the MR project in KF community
+* The resulting artifact cannot be re-used by other communities/users
```
Comment (lampajr) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md:110]:
```suggestion
* The resulting artifact cannot be re-used by other communities/users
* If one day upstream Google MLMD resolves to be platform/arch agnostic, changing back the MR python client to use the original ml-metadata could require extra work and effort
```

**Review by lampajr — DISMISSED** [review]
lgtm!

I just left another cons in `packaging:option2`, feel free to ignore it if you do not agree on that

**Comment by rareddy** [comment]
why sharding, can we not promote as a separate package? namespace collisions?

**Comment by tarilabs** [comment]
> why sharding

I'm not sure I understand which sharding is referenced here, please?

**Comment by rareddy** [comment]
>I'm not sure I understand which sharding is referenced here, please?

It is mentioned that MLMD libraries will be packaged along with the Model Registry libraries as uber jar. Wondering why?

**Comment by tarilabs** [comment]
> It is mentioned that MLMD libraries will be packaged along with the Model Registry libraries as uber jar. Wondering why?

That is one of the alternatives _considered_.

Then, based on SWOT/pro-cons analysis as detailed, the ADR **as a result recommend actually aiming** to:

1. soft-fork upstream repo, modify pip+bazel build, to produce the wheel (Alternative B)
ref: https://github.com/opendatahub-io/model-registry/pull/261/files#diff-1d7987bd8dfd381e577fb15c1a68e82a098d9e9bebe5e1a5a21c447b7d3ec7d8R114

2.  Option1 for Packaging of distribution, ie create a `ml-metadata-remote` package on PyPI based on the wheel produced from the step1 above
ref: https://github.com/opendatahub-io/model-registry/pull/261/files#diff-1d7987bd8dfd381e577fb15c1a68e82a098d9e9bebe5e1a5a21c447b7d3ec7d8R115

So we're NOT going to "uberjar" inside of MR python client.

Does this clarifies, please?

**Comment by rareddy** [comment]
yes, thanks

**Comment by tarilabs** [comment]
> yes, thanks

glad to hear it :) happy to hear other comments, or kindly approve if no other feedback when possible, please.

**Review by rareddy — DISMISSED** [review]
lgtm

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
+# Remote-only packaging of MLMD Python lib
+
+Authors: 
+- Matteo Mortari
+- Isabella do Amaral
+- and by extension the MR team
+
+## Context and Problem statement
+
+Google’s ML Metadata (MLMD) is a project composed of a cpp server, and a python client library. The cpp code is compiled and distributed only for specific architectures; the resulting binary executable of the MLMD server exposes a gRPC interface. The MLMD python wheel/library is not distributed as a pure library, but embeds the compiled cpp server as a binary executable inside of the python wheel, making the resulting MLMD python wheel as well [platform/architecture specific](https://pypi.org/project/ml-metadata/1.14.0/#files). The reason for embedding the cpp server binary executable in the python wheel of MLMD is to provide [a convenient method for running the server “locally”](https://www.tensorflow.org/tfx/guide/mlmd#metadata_storage_backends_and_store_connection_configuration) (without Docker etc) to the end-user of the python library, while still supporting the [option for gRPC connections](https://www.tensorflow.org/tfx/guide/mlmd#use_mlmd_with_a_remote_grpc_server) to remote MLMD servers.
```
Comment (antisaling) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md]: Can we simplify like so?

```suggestion
Google’s ML Metadata (MLMD) is a project composed of a C++ server, and a Python client library.
The server exposes a gRPC interface, and is only distributed for x86-64 architectures.
It is embedded in the client's wheel binary, providing an additional convenience [method for running the server locally (in memory)](https://www.tensorflow.org/tfx/guide/mlmd#metadata_storage_backends_and_store_connection_configuration), 
whilst also making it [architecture specific](https://pypi.org/project/ml-metadata/1.14.0/#files).
```

I also wouldn't touch on **platform** specific because as far as we're concerned they do compile for macs, just not for arm64 :(.

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+Our Model Registry (MR) [project](https://docs.google.com/document/d/1G-pjdGaS2kLELsB5kYk_D4AmH-fTfnCnJOhJ8xENjx0/edit?usp=sharing) is built on top of MLMD. The Go implementation interfaces with the MLMD server via gRPC, typically available as a Docker container. The [MR python client](https://github.com/opendatahub-io/model-registry/tree/main/clients/python#readme) wraps MLMD python library, especially for reusing types (Context, Artifact, etc) and for the gRPC connection methods.
```
Comment (antisaling) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md]: I don't think we need the final part.

```suggestion
The [Model Registry project](https://docs.google.com/document/d/1G-pjdGaS2kLELsB5kYk_D4AmH-fTfnCnJOhJ8xENjx0/edit?usp=sharing) (MR) is built on top of MLMD.
The Go implementation interfaces with the MLMD server via gRPC, typically available as a Docker container.
The [MR Python client](https://github.com/opendatahub-io/model-registry/tree/main/clients/python#readme) wraps the MLMD client.
```

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+The fact that the MLMD python wheel is platform/architecture specific poses some challenges, for example:
+
+* It makes **development of Model Registry** difficult for contributors using Apple-silicon Macs, or for contributors from ARM-based computers.
+* It also makes the **Model Registry python client** deployable only on the same specific platform/architecture where Google's MLMD wheel has been actually distributed for (e.g.: might be difficult to deploy MR client on ARM, due to the MLMD dependency not being natively available for ARM).
```
Comment (antisaling) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md]: Also think this could be simplified a little, wdyt?

```suggestion
As the MLMD client is architecture specific, so is the MR Python client, which **severely limits the targets it can run on**, as it only supports x86-64.
This **poses many challenges to contributors** using other CPU architectures, specially ARM, as that's become more prevalent in recent years.
```

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+Since the Model Registry python client does _not_ leverage as a capability the embedded MLMD server of the python wheel, this documents discusses options for a “soft-fork” and re-distribution of the MLMD python wheel to be only supporting remote gRPC connections while at the same time becoming a “pure library” and platform/architecture agnostic.
```
Comment (antisaling) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md]: Maybe we can keep platform/architecture agnostic here, but I'd rephrase it as such:

```suggestion
Since the Model Registry python client does _not_ leverage the embedded MLMD server's in-memory connection, we present options for a “soft-fork” and re-distribution of the MLMD client as a pure Python library, making it platform/architecture agnostic.
```

**Review by antisaling — COMMENTED** [review]
We have to consider that in alternative A we're only publishing a wheel. That has [several disadvantages](https://discuss.python.org/t/sdists-for-pure-python-projects/25191), so it definitely impacts our choice as well.
I'm just not very sure where it fits, so I will suggest a slight rearrangement on the final part:

- Start with the first solution you explored, then we discuss the uber-jar packaging option and then publishing on PyPI with the caveat above.
- Then we talk about the soft-fork alternative, which solves the packaging caveat, with the added pros.

Maybe it'd also be better to push aside some technicalities on the wheel packaging process as they're already well documented on the repos?

Let me know what you think :)

**Comment by tarilabs** [comment]
> Maybe it'd also be better to push aside some technicalities on the wheel packaging process as they're already well documented on the repos?

I prefer to keep it as they document the efforts in the document.

**Comment by tarilabs** [comment]
I have incorporated all comments and reworked slightly some sections following other feedback with b926169 .

I hope is good to merge now.

File: doc/remote_only_packaging_of_MLMD_Python_lib.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    * at this stage is not clear if any implications with uplifting the MR project in KF community
+* The resulting artifact cannot be re-used by other communities/users
+* If one day upstream Google MLMD resolves to be platform/arch agnostic, changing back the MR python client to use the original ml-metadata could require extra work and effort
+
+
+## Conclusion
+
+Based on analysis of the alternatives provided, we decided to further pursue:
+- the repackaging by **Alternative B** because makes it actually easier to demonstrate the steps and modifications required using as baseline the upstream repo.
+- the distribution by **Pacakging Option1** because it will make it easier to "revert" to the upstream `ml-metadata` dependency if upstream will publish for all architectues, OSes, etc. and as the pros outweight considered cons.
```
Comment (antisaling) [inline doc/remote_only_packaging_of_MLMD_Python_lib.md]:
```suggestion
- the distribution by **Packaging Option1** because it will make it easier to "revert" to the upstream `ml-metadata` dependency if upstream will publish for all architectures, OSes, etc. and as the pros outweight considered cons.
```

**Review by antisaling — DISMISSED** [review]
Minor correction. Otherwise LGTM!

File: doc/remote_only_packaging_of_MLMD_Python_lib.md (hunk 6 above)
Comment (tarilabs) [reply doc/remote_only_packaging_of_MLMD_Python_lib.md]: thanks for the catch!

**Review by lampajr — APPROVED** [review]

**Review by antisaling — APPROVED** [review]

**Review by danielezonca — APPROVED** [review]
lgtm
